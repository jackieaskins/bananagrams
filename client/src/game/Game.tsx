import { Suspense } from 'react';

import BoardWrapper from '../board/LazyBoardWrapper';
import LoadingContainer from '../loading/LoadingContainer';
import Overlay from '../overlay/Overlay';
import SelectedTile from '../tile/SelectedTile';

const Game = (): JSX.Element => (
  <div
    css={{
      width: '100vw',
      height: '100vh',
    }}
  >
    <Suspense fallback={<LoadingContainer size="large" />}>
      <BoardWrapper />
    </Suspense>
    <Overlay />
    <SelectedTile />
  </div>
);

export default Game;
