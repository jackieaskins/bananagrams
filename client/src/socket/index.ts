import io from 'socket.io-client';

const SOCKET_URL = 'http://localhost:5000';

export default process.env.NODE_ENV === 'development' ? io(SOCKET_URL) : io();
