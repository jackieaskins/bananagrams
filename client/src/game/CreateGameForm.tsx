import { Button, Form as AntForm, Input } from 'antd';
import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';

import Form, { SubmitFormFn } from '../form/Form';
import { GameInfo, GameLocationState } from '../games/types';
import { useSocket } from '../socket/SocketContext';

const FormItem = AntForm.Item;

type FormValues = {
  gameName: string;
  username: string;
};

type CreateGameFormProps = {
  isShortenedGame?: boolean;
};

const CreateGameForm = ({
  isShortenedGame = false,
}: CreateGameFormProps): JSX.Element => {
  const { socket } = useSocket();
  const { push } = useHistory();

  const handleSubmit: SubmitFormFn<FormValues> = useCallback(
    (setError) => ({ gameName, username }) => {
      socket.emit(
        'createGame',
        { gameName, username, isShortenedGame },
        (error: Error, gameInfo: GameInfo) => {
          if (error) {
            setError(error.message);
          } else {
            const locationState: GameLocationState = {
              isInGame: true,
              gameInfo,
            };

            push(`/game/${gameInfo.gameId}`, locationState);
          }
        }
      );
    },
    [isShortenedGame, push, socket]
  );

  return (
    <Form labelCol={{ span: 5 }} onSubmit={handleSubmit}>
      <FormItem label="Game name" name="gameName" rules={[{ required: true }]}>
        <Input />
      </FormItem>

      <FormItem label="Username" name="username" rules={[{ required: true }]}>
        <Input />
      </FormItem>

      <FormItem>
        <Button type="primary" htmlType="submit" block>
          Create game
        </Button>
      </FormItem>
    </Form>
  );
};

export default CreateGameForm;
