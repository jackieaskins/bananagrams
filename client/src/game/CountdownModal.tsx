import { Modal } from 'antd';
import { useMemo } from 'react';

import { getUserId } from '../socket';
import { useGameCountdown, useGamePlayers, useGameStatus } from './stateHooks';

const CountdownModal = (): JSX.Element => {
  const userId = getUserId();
  const gameStatus = useGameStatus();
  const countdown = useGameCountdown();
  const players = useGamePlayers();

  const topBanana = useMemo(() => {
    const winner = players.find(({ isTopBanana }) => isTopBanana);

    if (!winner) {
      return 'Someone';
    }

    return userId === winner.userId ? 'You' : winner.username;
  }, [players, userId]);

  const modalTitle = useMemo(
    () => (gameStatus === 'STARTING' ? 'Game starting' : 'Game over'),
    [gameStatus]
  );

  const modalText = useMemo(() => {
    const countdownText = `${countdown} second${countdown === 1 ? '' : 's'}`;

    if (gameStatus === 'STARTING') {
      return `Everyone is ready, the game will start in ${countdownText}`;
    }

    return `${topBanana} won! The game will end in ${countdownText}`;
  }, [gameStatus, countdown, topBanana]);

  return (
    <Modal
      closable={false}
      footer={null}
      title={modalTitle}
      visible={['STARTING', 'ENDING'].includes(gameStatus)}
    >
      {modalText}
    </Modal>
  );
};

export default CountdownModal;
