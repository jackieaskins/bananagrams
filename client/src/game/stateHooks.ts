import { useRecoilCallback, useRecoilValue, useSetRecoilState } from 'recoil';

import {
  Board,
  BoardPosition,
  BoardSquare,
  getSquareId,
} from '../boards/types';
import { initializeState } from '../game/state';
import { GameInfo, GameStatus } from '../games/types';
import { Hand } from '../hands/types';
import { Player } from '../players/types';
import { SetState } from '../state/types';
import { Tile } from '../tiles/types';

const {
  statusState,
  countdownState,
  nameState,
  bunchState,
  bunchCountState,
  playersState,
  currentPlayerState,
  handsState,
  currentHandState,
  boardsState,
  currentBoardState,
  currentBoardSquaresState,
} = initializeState();

export const useGameStatus = (): GameStatus => useRecoilValue(statusState);

export const useGameCountdown = (): number => useRecoilValue(countdownState);

export const useGameName = (): string => useRecoilValue(nameState);

export const useGameBunch = (): Tile[] => useRecoilValue(bunchState);
export const useGameBunchCount = (): number => useRecoilValue(bunchCountState);

export const useGamePlayers = (): Player[] => useRecoilValue(playersState);
export const useCurrentPlayer = (): Player | null =>
  useRecoilValue(currentPlayerState);

export const useGameHands = (): Record<string, Hand> =>
  useRecoilValue(handsState);

export const useCurrentHand = (): Hand => useRecoilValue(currentHandState);
export const useSetCurrentHand = (): SetState<Hand> =>
  useSetRecoilState(currentHandState);

export const useGameBoards = (): Record<string, Board> =>
  useRecoilValue(boardsState);
export const useCurrentBoard = (): Board => useRecoilValue(currentBoardState);
export const useCurrentBoardSquare = (
  position: BoardPosition
): BoardSquare | null =>
  useRecoilValue(currentBoardSquaresState(getSquareId(position)));

export const useResetCurrentBoard = (): (() => void) =>
  useRecoilCallback(({ reset, snapshot }) => async () => {
    const oldBoard = await snapshot.getPromise(currentBoardState);

    reset(currentBoardState);
    Object.keys(oldBoard).forEach((id) => {
      reset(currentBoardSquaresState(id));
    });
  });

export const useSetCurrentBoard = (): ((board: Board) => void) =>
  useRecoilCallback(
    ({ set }) => (board) => {
      set(currentBoardState, board);
      Object.entries(board).forEach(([id, square]) => {
        set(currentBoardSquaresState(id), (currSquare) => {
          if (
            currSquare &&
            square &&
            currSquare.tile.id === square.tile.id &&
            currSquare.tile.letter === square.tile.letter &&
            currSquare.validationStatus === square.validationStatus
          ) {
            return currSquare;
          }

          return square;
        });
      });
    },
    []
  );

export const useResetGameState = (): (() => void) =>
  useRecoilCallback(
    ({ reset }) => () => {
      reset(statusState);
      reset(countdownState);
      reset(nameState);
      reset(bunchState);
      reset(playersState);
      reset(handsState);
      reset(boardsState);
    },
    []
  );

export const useUpdateGameState = (): ((gameInfo: GameInfo) => void) =>
  useRecoilCallback(
    ({ set }) => (gameInfo) => {
      set(statusState, gameInfo.status);
      set(countdownState, gameInfo.countdown);
      set(nameState, gameInfo.gameName);
      set(bunchState, gameInfo.bunch);
      set(playersState, gameInfo.players);
      set(handsState, gameInfo.hands);
      set(boardsState, gameInfo.boards);
    },
    []
  );
