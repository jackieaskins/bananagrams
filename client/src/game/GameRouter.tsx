import { useEffect } from 'react';

import { GameInfo } from '../games/types';
import { addGameInfoListener, removeGameInfoListener } from '../socket';
import Game from './Game';
import WaitingRoom from './WaitingRoom';
import { useGameStatus, useUpdateGameState } from './stateHooks';

type GameRouterProps = {
  initialGameInfo: GameInfo;
};

const GameRouter = ({ initialGameInfo }: GameRouterProps): JSX.Element => {
  const updateGameState = useUpdateGameState();
  const gameStatus = useGameStatus();
  const isGameInProgress = gameStatus === 'IN_PROGRESS';

  useEffect(() => {
    updateGameState(initialGameInfo);

    addGameInfoListener((info: GameInfo) => {
      updateGameState(info);
    });

    return (): void => {
      removeGameInfoListener();
    };
  }, [initialGameInfo, updateGameState]);

  if (isGameInProgress) {
    return <Game />;
  }

  return <WaitingRoom />;
};

export default GameRouter;
