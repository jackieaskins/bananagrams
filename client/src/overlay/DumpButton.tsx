import { Button } from 'antd';
import { memo, useCallback, useMemo } from 'react';
import { useRecoilState } from 'recoil';

import { useGameBunchCount } from '../game/stateHooks';
import { dump } from '../socket';
import { selectedTileState } from '../tile/selectedTileState';

const EXCHANGE_COUNT = 3;

const DumpButton = (): JSX.Element => {
  const [selectedTile, setSelectedTile] = useRecoilState(selectedTileState);
  const bunchCount = useGameBunchCount();

  const isDisabled = useMemo(
    () => !selectedTile || bunchCount < EXCHANGE_COUNT,
    [selectedTile, bunchCount]
  );

  const handleClick = useCallback(() => {
    if (selectedTile) {
      dump({
        boardPosition: selectedTile.boardPosition,
        tileId: selectedTile.tile.id,
      });
      setSelectedTile(null);
    }
  }, [selectedTile, setSelectedTile]);

  return (
    <Button
      className="no-drop"
      block
      disabled={isDisabled}
      onClick={handleClick}
    >
      Dump
    </Button>
  );
};

export default memo(DumpButton);
