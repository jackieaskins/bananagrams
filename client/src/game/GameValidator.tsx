import { useEffect, useState } from 'react';
import { Redirect, useHistory, useLocation, useParams } from 'react-router-dom';

import { GameLocationState } from '../games/types';
import { leaveGame } from '../socket';
import CountdownModal from './CountdownModal';
import GameRouter from './GameRouter';

const GameValidator = (): JSX.Element => {
  const { replace } = useHistory();
  const { pathname, state } = useLocation<GameLocationState>();
  const { gameId } = useParams<{ gameId: string }>();

  const [gameInfo] = useState(state?.gameInfo);

  useEffect(() => {
    if (gameInfo) {
      replace(pathname);

      return () => {
        leaveGame({ gameId });
      };
    }

    return;
  }, [gameId, gameInfo, pathname, replace]);

  if (gameInfo) {
    return (
      <>
        <GameRouter initialGameInfo={gameInfo} />
        <CountdownModal />
      </>
    );
  }

  return <Redirect to={`/game/${gameId}/join`} />;
};

export default GameValidator;
