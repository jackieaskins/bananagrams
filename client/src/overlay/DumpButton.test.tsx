import { Button } from 'antd';
import { shallow } from 'enzyme';
import { useRecoilState } from 'recoil';

import { useGameBunchCount } from '../game/stateHooks';
import { dump } from '../socket';
import DumpButton from './DumpButton';

const mockSetSelectedTile = jest.fn();
const mockUseRecoilState = useRecoilState as jest.Mock;
jest.mock('recoil', () => ({
  useRecoilState: jest.fn(),
}));

const mockUseGameBunchCount = useGameBunchCount as jest.Mock;
jest.mock('../game/stateHooks', () => ({
  useGameBunchCount: jest.fn(),
}));

const mockDump = dump as jest.Mock;
jest.mock('../socket', () => ({
  dump: jest.fn(),
}));

jest.mock('../tile/selectedTileState', () => ({
  selectedTileState: null,
}));

const tile = { id: 'A1', letter: 'A' };
const selectedTile = { tile, boardPosition: null };
describe('<DumpButton />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<DumpButton {...propOverrides} />);

  beforeEach(() => {
    mockUseRecoilState.mockImplementation((init) => [
      init,
      mockSetSelectedTile,
    ]);
    mockUseGameBunchCount.mockReturnValue(0);
  });

  it('renders button', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  describe('disabled state', () => {
    const isButtonDisabled = () =>
      renderComponent().find(Button).props().disabled;

    it('is disabled if no selected tile', () => {
      mockUseRecoilState.mockReturnValue([null, mockSetSelectedTile]);
      mockUseGameBunchCount.mockReturnValue(4);

      expect(isButtonDisabled()).toEqual(true);
    });

    it('is disabled if bunch has less than exchange count', () => {
      mockUseRecoilState.mockReturnValue([selectedTile, mockSetSelectedTile]);
      mockUseGameBunchCount.mockReturnValue(0);

      expect(isButtonDisabled()).toEqual(true);
    });

    it('is enabled if selected tile and enough tiles left in bunch', () => {
      mockUseRecoilState.mockReturnValue([selectedTile, mockSetSelectedTile]);
      mockUseGameBunchCount.mockReturnValue(4);

      expect(isButtonDisabled()).toEqual(false);
    });
  });

  describe('button click', () => {
    const clickButton = () => renderComponent().find(Button).simulate('click');

    it('does not dump if no selected tile', () => {
      mockUseRecoilState.mockReturnValue([null, mockSetSelectedTile]);
      clickButton();

      expect(mockDump).not.toHaveBeenCalled();
    });

    it('dumps if selected tile', () => {
      mockUseRecoilState.mockReturnValue([selectedTile, mockSetSelectedTile]);
      clickButton();

      expect(mockDump).toHaveBeenCalledWith({
        boardPosition: selectedTile.boardPosition,
        tileId: selectedTile.tile.id,
      });
    });

    it('clears selected tile if selected tile', () => {
      mockUseRecoilState.mockReturnValue([selectedTile, mockSetSelectedTile]);
      clickButton();

      expect(mockSetSelectedTile).toHaveBeenCalledWith(null);
    });
  });
});
