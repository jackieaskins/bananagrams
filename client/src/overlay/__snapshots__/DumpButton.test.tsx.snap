// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DumpButton /> renders button 1`] = `
<Tooltip
  arrowPointAtCenter={false}
  autoAdjustOverflow={true}
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  placement="bottom"
  title="Not enough tiles remaining in bunch to dump"
  transitionName="zoom-big-fast"
>
  <Button
    block={true}
    disabled={true}
    onClick={[Function]}
  >
    Dump
  </Button>
</Tooltip>
`;

exports[`<DumpButton /> tooltipText shows explanation tooltip otherwise 1`] = `"Exchange the selected tile for 3 from the bunch"`;

exports[`<DumpButton /> tooltipText shows no selected tile if no selected tile 1`] = `"A tile must be selected to dump"`;

exports[`<DumpButton /> tooltipText shows not enough tils if less than exchange count 1`] = `"Not enough tiles remaining in bunch to dump"`;
